owltorrent-network/src/main/java/edu/rice/owltorrent/network/TorrentManager.java

```diff
@@ -0,0 +1,199 @@
+import edu.rice.owltorrent.common.entity.Peer;
+import edu.rice.owltorrent.common.entity.Torrent;
+import edu.rice.owltorrent.common.util.AtomicHashableInteger;
+import edu.rice.owltorrent.network.messages.PayloadlessMessage;
+import edu.rice.owltorrent.network.messages.PieceActionMessage;
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.extern.log4j.Log4j2;
+
+/**
+ * Manager that takes care of everything related to a Torrent. The TorrentManager will manage the
+ * peer connections and ask for missing file pieces.
+ *
+ * @author Lorraine Lyu, Max Yu
+ */
+@Log4j2(topic = "general")
+public class TorrentManager implements Runnable, AutoCloseable {
+
+  private static final int DEFAULT_BLOCK_NUM = 2;
+
+  private static final int BLOCK_NOT_STARTED = 0;
+  private static final int BLOCK_IN_PROGRESS = 1;
+  private static final int BLOCK_DONE = 2;
+
+  private final Set<Integer> completedPieces;
+  private final Map<Integer, PieceStatus> uncompletedPieces;
+  private final Set<Integer> notStartedPieces;
+
+  private final int totalPieces;
+
+  private final Map<Peer, PeerConnector> peers;
+  private final NetworkToStorageAdapter networkStorageAdapter;
+  @Getter private final Torrent torrent;
+
+  public TorrentManager(Torrent file, NetworkToStorageAdapter adapter) {
+    this.torrent = file;
+    this.networkStorageAdapter = adapter;
+    this.completedPieces = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    this.uncompletedPieces = new ConcurrentHashMap<>();
+    this.notStartedPieces = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    this.totalPieces = torrent.getPieces().size();
+
+    for (int idx = 0; idx < totalPieces; idx++) {
+      this.notStartedPieces.add(idx);
+    }
+
+    this.peers = new ConcurrentHashMap<>();
+
+    // initPeers();
+  }
+
+  private void initPeers(List<Peer> peerList) {
+    for (Peer peer : peerList) {
+      try {
+        PeerConnector connector =
+            SocketConnector.makeInitialConnection(peer, this, networkStorageAdapter);
+        addPeer(connector, peer);
+        // TODO: revise
+        connector.writeMessage(new PayloadlessMessage(PeerMessage.MessageType.INTERESTED));
+      } catch (IOException e) {
+        log.error(String.format("Error connecting peer id=%s", peer.getPeerID().toString()));
+      }
+    }
+  }
+
+  public void startDownloadingAsynchronously() {
+    // TODO: make thread pool scheduler
+    new Thread(this).start();
+  }
+
+  public float getProgressPercent() {
+    return completedPieces.size() * 1.0f / totalPieces;
+  }
+
+  // For handshake listener
+  // Later if get updated peerlist from tracker.
+  public void addPeer(PeerConnector connector, Peer peer) {
+    this.peers.put(peer, connector);
+  }
+
+  @Override
+  public void close() throws Exception {
```

Comment:
throws Exception isn't good, make them specific, or add a specific custom exception to our exception utils class (I created it in the other pr)

Reply:
This is actually the signature for Java AutoClosable.

Reply:
okay fair enough